x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardFollowTaskReplicationTests.java

```diff
@@ -0,0 +1,218 @@
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.ccr.action;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.replication.TransportWriteAction;
+import org.elasticsearch.cluster.metadata.IndexMetaData;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.index.IndexSettings;
+import org.elasticsearch.index.engine.Engine.Operation.Origin;
+import org.elasticsearch.index.engine.EngineFactory;
+import org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.index.translog.Translog;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.xpack.ccr.CcrSettings;
+import org.elasticsearch.xpack.ccr.action.bulk.BulkShardOperationsRequest;
+import org.elasticsearch.xpack.ccr.action.bulk.BulkShardOperationsResponse;
+import org.elasticsearch.xpack.ccr.action.bulk.TransportBulkShardOperationsAction;
+import org.elasticsearch.xpack.ccr.index.engine.FollowingEngineFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.LongConsumer;
+
+public class ShardFollowTaskReplicationTests extends ESIndexLevelReplicationTestCase {
+
+    public void testSimpleCcrReplication() throws Exception {
+        try (ReplicationGroup leaderGroup = createGroup(randomInt(2))) {
+            leaderGroup.startAll();
+            int docCount = leaderGroup.appendDocs(randomInt(64));
+            leaderGroup.assertAllEqual(docCount);
+            try (ReplicationGroup followerGroup = createFollowGroup(randomInt(2))) {
+                followerGroup.startAll();
+                ShardFollowNodeTask shardFollowTask = createShardFollowTask(leaderGroup, followerGroup);
+                shardFollowTask.start(followerGroup.getPrimary().getGlobalCheckpoint());
+                docCount += leaderGroup.appendDocs(randomInt(128));
+                leaderGroup.syncGlobalCheckpoint();
+
+                leaderGroup.assertAllEqual(docCount);
+                int expectedCount = docCount;
+                assertBusy(() -> followerGroup.assertAllEqual(expectedCount));
+                shardFollowTask.markAsCompleted();
+            }
+        }
+    }
+
+    public void testFailLeaderReplicaShard() throws Exception {
+        try (ReplicationGroup leaderGroup = createGroup(1 + randomInt(1))) {
+            leaderGroup.startAll();
+            try (ReplicationGroup followerGroup = createFollowGroup(randomInt(2))) {
```

Comment:
Same comment as above. I think we can put both leaderGroup and followerGroup into a single try-clause.